core/shared/src/main/scala/zio/stm/TMap.scala

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.stm
+
+/**
+ * Transactional map implemented on top of [[TRef]] and [[TArray]]. Resolves
+ * conflicts via chaining.
+ *
+ * Caution: doesn't provide stack-safety guarantees.
+ */
+class TMap[K, V] private (private val tBuckets: TRef[TArray[List[(K, V)]]], private val tCapacity: TRef[Int]) { self =>
+  import TMap.{ allocate, LoadFactor }
+
+  /**
+   * Tests whether or not map contains a key.
+   */
+  final def contains[E](k: K): STM[E, Boolean] =
+    get(k).map(_.isDefined)
+
+  /**
+   * Removes binding for given key.
+   */
+  final def delete[E](k: K): STM[E, Unit] =
+    for {
+      buckets <- tBuckets.get
+      idx     <- indexOf(k)
+      _       <- buckets.update(idx, _.filterNot(_._1 == k))
+    } yield ()
+
+  /**
+   * Atomically folds using pure function.
+   */
+  final def fold[A](zero: A)(op: (A, (K, V)) => A): STM[Nothing, A] =
+    for {
+      buckets <- tBuckets.get
+      res     <- buckets.fold(zero)((acc, chain) => chain.foldLeft(acc)(op))
+    } yield res
+
+  /**
+   * Atomically folds using effectful function.
+   */
+  final def foldM[A, E](zero: A)(op: (A, (K, V)) => STM[E, A]): STM[E, A] = {
+    def loopM(acc: STM[E, A], remaining: List[(K, V)]): STM[E, A] =
+      remaining match {
+        case Nil          => acc
+        case head :: tail => loopM(acc.flatMap(op(_, head)), tail)
+      }
+
+    for {
+      buckets <- tBuckets.get
+      res     <- buckets.foldM(zero)((acc, chain) => loopM(STM.succeed(acc), chain))
+    } yield res
+  }
+
+  /**
+   * Atomically performs side-effect for each binding present in map.
+   */
+  final def foreach[E](f: ((K, V)) => STM[E, Unit]): STM[E, Unit] =
+    self.foldM(())((_, kv) => f(kv))
+
+  /**
+   * Retrieves value associated with given key.
+   */
+  final def get[E](k: K): STM[E, Option[V]] =
+    for {
+      buckets <- tBuckets.get
+      idx     <- indexOf(k)
+      bucket  <- buckets(idx)
+    } yield bucket.find(_._1 == k).map(_._2)
+
+  /**
+   * Retrieves value associated with given key or default value, in case the
+   * key isn't present.
+   */
+  final def getOrElse[E](k: K, default: => V): STM[E, V] =
+    get(k).map(_.getOrElse(default))
+
+  /**
+   * Creates new [[TMap]] by mapping all bindings using pure function.
+   */
+  final def map[K2, V2](f: ((K, V)) => (K2, V2)): STM[Nothing, TMap[K2, V2]] =
```

Comment:
Personally I find these pure methods confusing. I would delete map and mapM, leaving only transform and transformM

Reply:
Agreed, done.